test/unit/code_writer_test.rb

```diff
@@ -423,4 +423,24 @@ def test_write_push_pop_writes_another_C_POP_temp_to_output
 
     assert_equal(expected, output.string)
   end
+
+  def test_write_push_pop_writes_C_PUSH_local
+    output = StringIO.new
+    code_writer = CodeWriter.new(output)
+    code_writer.write_push_pop(:C_PUSH, "local", 0)
+    expected = <<~EOF
+      @0
```

Comment:
I've checked to make sure that this code works on the `CPUEmulator` and it works the way I think it should. Does it look correct to you? Is there anything that can be improved? @tomstuart

Reply:
Yeah, it looks right to me! My confidence is significantly increased by your approach of doing surgery on the specific part of the `push` implementation which needed to change (i.e. loading the value into `D`) without touching the rest, because we already know the other parts work. 👍🏻 

One organisational improvement to consider is that we have a couple of places where we’re assigning a local variable (currently called `final_memory_address`) inside a conditional, e.g.:

```ruby
if segment == # …
  whatever = # …
elsif segment == # …
  whatever = # …
else
  whatever = # …
end
```

(This would be clearer as a `case segment`, but ignoring that for now…)

In this situation we can always lift the assignment outside of the conditional, which reduces duplication and makes it more obvious what the overall effect of the code is:

```ruby
whatever =
  if segment == # …
    # …
  elsif segment == # …
    # …
  else
    # …
  end
```

The indented conditional is itself somewhat awkward, which can be an indication that we should just extract a (private?) helper method for the conditional to live inside:

```ruby
whatever = figure_out_whatever(segment)

# …

private

def figure_out_whatever(segment)
  if segment == # …
    # …
  elsif segment == # …
    # …
  else
    # …
  end
end
```

This creates the opportunity to give the helper a meaningful name (hopefully better than “`#figure_out_whatever`”) and slightly raises the level of abstraction of the calling code by hiding the particular detail of how the value is calculated. Also, if the local variable is actually only used once, we might be able to get rid of it entirely and just call the helper method inline at the place where its value is needed.